File: Year 2/Semester 2/MPP/C#Socket/Persistency/EmployeeRepository.cs

using log4net;
using System;
using System.Collections.Generic;
using System.Data.Common;
using System.Data.SQLite;
using System.Data;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Model;
using Persistency.utils;

namespace Persistency
{
    public class EmployeeRepository : IEmployeeRepository
    {
        private static readonly ILog log = LogManager.GetLogger("EmployeeRepository");
        private IDictionary<string, string> props;
        public EmployeeRepository(IDictionary<string, string> props)
        {
            log.Info("Creating EmployeeRepository");
            this.props = props;
        }

        public void add(Employee entity)
        {
            log.Info("Entering add function");
            var connection = (SQLiteConnection)DBUtils.GetConnection(props);
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "insert into employees values(@id,@first_name,@last_name,@username,@password)";
                var paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = null;

                var paramFirstName = command.CreateParameter();
                paramFirstName.ParameterName = "@first_name";
                paramFirstName.Value = entity.First_name;

                var paramLastName = command.CreateParameter();
                paramLastName.ParameterName = "@last_name";
                paramLastName.Value = entity.Last_name;

                var paramUsername = command.CreateParameter();
                paramUsername.ParameterName = "@username";
                paramUsername.Value = entity.Username;

                var paramPassword = command.CreateParameter();
                paramPassword.ParameterName = "@password";
                paramPassword.Value = entity.Password;

                command.Parameters.Add(paramId);
                command.Parameters.Add(paramFirstName);
                command.Parameters.Add(paramLastName);
                command.Parameters.Add(paramUsername);
                command.Parameters.Add(paramPassword);
                var result = command.ExecuteNonQuery();

                if (result == 0)
                    log.Error("No entities were added in Employee table");
                else
                    log.InfoFormat("{0} were added in Employee table", result);
            }
        }

        public Employee findById(int id)
        {
            log.InfoFormat("Entering findById with value {0}", id);
            var connection = (SQLiteConnection)DBUtils.GetConnection(props);
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select employee_id,first_name,last_name,username,password from emplyoees where id=@id";
                IDbDataParameter paramId = command.CreateParameter();
                paramId.ParameterName = "@id";
                paramId.Value = id;
                command.Parameters.Add(paramId);
                using (var reader = command.ExecuteReader())
                {
                    Employee employee = extractEntity(reader);
                    log.InfoFormat("Exiting findById with value {0}", employee);
                    return employee;
                }
            }
            log.InfoFormat("Exiting findById with value {0}", null);
            return null;

        }
        private Employee extractEntity(DbDataReader reader)
        {
            int id = reader.GetInt32(0);
            string first_name = reader.GetString(1);
            string last_name = reader.GetString(2);
            string username = reader.GetString(3);
            string password = reader.GetString(4);
            Employee employee = new Employee(first_name, last_name, username, password);
            employee.Id = id;
            return employee;
        }
        public List<Employee> GetAll()
        {
            var connection = (SQLiteConnection)DBUtils.GetConnection(props);
            List<Employee> employees = new List<Employee>();
            using (var command = connection.CreateCommand())
            {
                command.CommandText = "select employee_id,first_name,last_name,username,password from employees";
                using (var reader = command.ExecuteReader())
                {
                    while (reader.Read())
                    {
                        Employee employee = extractEntity(reader);
                        employees.Add(employee);
                    }
                }
            }
            return employees;

        }
    }
}

File: Year 2/Semester 2/MPP/RestServices/Java REST/AirlinePersistence/src/main/java/airline/persistence/EmployeeRepository.java
package airline.persistence;



import airline.model.Employee;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class EmployeeRepository implements IEmployeeRepository {
    private static final Logger logger= LogManager.getLogger();
    private JdbcUtils dbUtils;

    public EmployeeRepository(Properties properties) {
        dbUtils=new JdbcUtils(properties);
    }

    public Iterable<Employee> getAll() {
        logger.traceEntry();
        Connection connection= dbUtils.getConnection();
        List<Employee> employeeList=new ArrayList<>();
        try(PreparedStatement ps=connection.prepareStatement("select * from employees")) {
            try (ResultSet resultSet = ps.executeQuery()) {
                while (resultSet.next()) {

                    Employee employee=extractEntity(resultSet);
                    employeeList.add(employee);
                }
            }
        }
        catch(SQLException ex) {
            logger.error(ex);
        }
        logger.traceExit();
        return employeeList;
    }

    public Employee getEmployee(String username, String password){
        logger.traceEntry();
        Connection connection= dbUtils.getConnection();
        try(PreparedStatement ps=connection.prepareStatement("select * from employees where username=? and password=?")){
            ps.setString(1,username);
            ps.setString(2,password);
            try(ResultSet resultSet=ps.executeQuery()){
                resultSet.next();
                return extractEntity(resultSet);
            }
        }
        catch (SQLException ex){
            logger.error(ex);
        }
        return null;
    }
    private Employee extractEntity(ResultSet resultSet) throws SQLException {
        int id = resultSet.getInt("employee_id");
        String first_name = resultSet.getString("first_name");
        String last_name = resultSet.getString("last_name");
        String username = resultSet.getString("username");
        String password = resultSet.getString("password");
        Employee employee = new Employee(first_name, last_name, username, password);
        employee.setId(id);
        return employee;
    }


    public Employee findById(int id) {
        logger.traceEntry();
        Connection connection= dbUtils.getConnection();
        try(PreparedStatement ps=connection.prepareStatement("select * from employees where id=?")){
            ps.setInt(1,id);
            try(ResultSet resultSet=ps.executeQuery()){
                resultSet.next();
                return extractEntity(resultSet);
            }
        }
        catch (SQLException ex){
            logger.error(ex);
        }

        logger.traceExit();
        return null;
    }


    public void add(Employee entity) {
         logger.traceEntry();
         Connection connection= dbUtils.getConnection();
         try(PreparedStatement ps=connection.prepareStatement("insert into employees(first_name,last_name,username,password) values (?,?,?,?)")){
             ps.setString(1,entity.getFirst_name());
             ps.setString(2,entity.getLast_name());
             ps.setString(3,entity.getUsername());
             ps.setString(4,entity.getPassword());
             int result=ps.executeUpdate();
             logger.trace("Added {} instances",result);
         }
         catch (SQLException ex){
          logger.error(ex);
         }
         logger.traceExit();
    }
}
